# Random Forest vs Extra Trees

## 1. Overview
Scikit-Learn is packed with amazing algoritms which serves a wide variety of purposes. Needless to say one algoritm does not fit all datasets. Each algoritm comes with its pros and cons and our role is to keep experimenting and pick the algorithm which serve our needs.
In this notebook, we will look at a special case of the famous Random Forest ensemble called Extremely Randomized Trees or Extra Trees and will see how can this ensemble become helpful in our projects. We will also compare it with the traditional Random Forest in terms of computational complexity. To get started lets first look at a basic Decision Tree.

<img src="https://i.stack.imgur.com/Q18mk.png">

## 2. Decision Tree
As the name suggest, Decision Tree is a tree with levels (or a depth), each level has nodes and each node takes a decision based on a certain threshold. Each node splits the dataset into two which helps in classifying any instance into categories and collectively into target classes. Look at the following figure for a better understanding (<a href="https://towardsdatascience.com/an-introduction-to-decision-trees-with-python-and-scikit-learn-1a5ba6fc204f">Source</a>)

<img src="https://miro.medium.com/max/1400/1*fGX0_gacojVa6-njlCrWZw.png" width=600 height=400/>



Here, the first node divides the dataset by sex, at the second level the dataset is divided by the features Age and Pclass and so on. The important thing to note here is that the threshold to divide the tree at a particular node is selected by the algorithm on its own. It calculates this threshold by analysing the feature and threshold that will give the least gini impurity. In simple terms, if we have a dataset which classifies students into pass or fail and we have only one feature i.e Marks, the decision tree will find the passing score(threshold) on its own and train a model. The next time we enter a student's marks, the algorithm classifies the student into pass or fail using the computed passing score.

## 3. Random Forest
Going from Decision Tree to Random Forest is simple. Instead of a single tree on an empty land(let's say apple tree), imagine a forest with 1000s of trees(but not every tree is an apple tree). Each tree bear a slightly different fruit with different taste, colour, or altogether a different fruit. This variety usually makes the harvest of the forest much more high-yielding. 

The Decision Tree takes the whole dataset and creates a tree. While Random Forest produces many trees but each tree is not shown the entire training dataset, it is shown only a part of the dataset and then it predicts the class or value. Later, all predicted values of all trees are cumulated to make the final prediction. 

The random partial data is shown to each tree to bring diversity. Imagine if all the trees are shown the entire dataset, then each tree will bear the same identical apple fruit, which will add no additional value than just using a single Decision Tree algoritm. Look at the following image for better understanding (<a href="https://towardsdatascience.com/random-forest-classification-and-its-implementation-d5d840dbead0">Source</a>).

<img src="https://miro.medium.com/max/1148/0*a8KgF1IINziv7KIQ.png" />

## 4. Extra Trees
The "Random" in Random Forest is bceause we are using a random subset of the dataset. But what if instead of choosing the best possible threshold for each tree at each node, we simply choose a random threshold too. If you remember the example taken in Decision Trees, imagine the algorithm is not bothered to compute the passing score in the first attempt, instead it takes a random marks value which divides the dataset(let that value be 90). Now at the first level, the students are divided into two categories (greater than and less than 90). If a student scores more than 90 he will be declared as pass, but if he scores less than 90, then the algorithm will find another random threshold (between 0 and 90) and categorize the remaining instances further. This algorithm too will eventually create an accurate model but it will have far more number of levels and nodes than a Decision Tree. However, the random nature of choosing the threshold value will make it much more faster.

<img src="https://www.researchgate.net/profile/Navoneel-Chakrabarty/publication/341967355/figure/fig1/AS:901875410948097@1592035262515/Visual-Representation-of-Extra-Trees-Classifier.ppm">

## 5. Implementation

Lets create our Extra Trees model!  
### 5.1. Importing dataset 
We will be using breast cancer data present in Scikit-Learn datasets. This dataset has 30 features and 569 instances.

In [1]:
import pandas as pd
from sklearn.datasets import load_breast_cancer

In [2]:
data = load_breast_cancer()

In [3]:
data_X = pd.DataFrame(data["data"], columns=data['feature_names'])
data_X

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst radius  worst texture  \
0                   0.07871  ...        25.380          17.33   
1                   0.05667  ...        24.990          23.41   
2                   0.05999  ...        23.570          25.53   
3                   0.09744  ...        14.910          26.50   
4                   0.05883  ...        22.540          16.67   
..                      ...  ...           ...            ...   
564                 0.05623  ...        25.450          26.40   
565                 0.05533  ...        23.690          38.25   
566                 0.05648  ...        18.980          34.12   
567                 0.07016  ...        25.740          39.42   
568                 0.05884  ...         9.456          30.37   

     worst perimeter  worst area  worst smoothness  worst compactness  \
0             184.60      2019.0           0.16220            0.66560   
1             158.80      1956.0           0.12380            0.18660   
2             152.50      1709.0           0.14440            0.42450   
3              98.87       567.7           0.20980            0.86630   
4             152.20      1575.0           0.13740            0.20500   
..               ...         ...               ...                ...   
564           166.10      2027.0           0.14100            0.21130   
565           155.00      1731.0           0.11660            0.19220   
566           126.70      1124.0           0.11390            0.30940   
567           184.60      1821.0           0.16500            0.86810   
568            59.16       268.6           0.08996            0.06444   

     worst concavity  worst concave points  worst symmetry  \
0             0.7119                0.2654          0.4601   
1             0.2416                0.1860          0.2750   
2             0.4504                0.2430          0.3613   
3             0.6869                0.2575          0.6638   
4             0.4000                0.1625          0.2364   
..               ...                   ...             ...   
564           0.4107                0.2216         

In [4]:
data_Y = data["target"]
data_Y[:30]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0])

In [5]:
data_X.shape

(569, 30)

In [6]:
data_Y.shape

(569,)

### 5.2. Creating a basic Extra Trees Classifier

In [7]:
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.model_selection import cross_val_score

We will use 5-fold cross validation and take its average to compute the accuracy of model.

In [8]:
etc = ExtraTreesClassifier(random_state=0)
cv_score = cross_val_score(etc, data_X, data_Y, cv=5).mean()
print("The avergae Cross validation score is ", cv_score)

The avergae Cross validation score is  0.9701288619779538


Wow! Without any hyperparameter tuning we achieved a validation score of 97% The model is working pretty well on its own, still lets look at some parameters that we can tweak.

## 6. Hypertuning

Here are some of the popular tuning parameters provided by Scikit-Learn :- 
1. <b>n_estimators</b> : The number of trees in a forest. Default = 100
2. <b>max_depth</b> : The maximum depth of the forest. Default = None i.e the tree will continue to expand untill all nodes are pure.
3. <b>min_samples_split</b> : The minimum number of samples required to split a node. Default = 2
4. <b>min_samples_leaf</b> : The minimum number of samples required to be at leaf node. This parameter becomes important specially if max_depth=None. Default = 1.

Check out other parameters in Scikit-Learn's documentation(<a href="https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.ExtraTreesClassifier.html?highlight=extra#sklearn.ensemble.ExtraTreesClassifier">here</a>).

Here, we will hypertune n_estimators and max_depth by using GridSerchCV

In [9]:
from sklearn.model_selection import GridSearchCV

In [13]:
parameters = {'n_estimators':[50, 100, 500, 1000], 'max_depth':[5, 10, 50, 100, 500]}
etc = ExtraTreesClassifier()
clf = GridSearchCV(etc, parameters, cv=5)
clf.fit(data_X, data_Y)

GridSearchCV(cv=5, error_score=nan,
             estimator=ExtraTreesClassifier(bootstrap=False, ccp_alpha=0.0,
                                            class_weight=None, criterion='gini',
                                            max_depth=None, max_features='auto',
                                            max_leaf_nodes=None,
                                            max_samples=None,
                                            min_impurity_decrease=0.0,
                                            min_impurity_split=None,
                                            min_samples_leaf=1,
                                            min_samples_split=2,
                                            min_weight_fraction_leaf=0.0,
                                            n_estimators=100, n_jobs=None,
                                            oob_score=False, random_state=None,
                                            verbose=0, warm_start=False),
             iid='deprecate

In [14]:
clf.best_estimator_

ExtraTreesClassifier(bootstrap=False, ccp_alpha=0.0, class_weight=None,
                     criterion='gini', max_depth=500, max_features='auto',
                     max_leaf_nodes=None, max_samples=None,
                     min_impurity_decrease=0.0, min_impurity_split=None,
                     min_samples_leaf=1, min_samples_split=2,
                     min_weight_fraction_leaf=0.0, n_estimators=100,
                     n_jobs=None, oob_score=False, random_state=None, verbose=0,
                     warm_start=False)

In [15]:
clf.best_score_

0.9736376339077782

Turns out that max_depth=500 and n_estimators=50 gives the best score of 97.3% which is a slight improvement over default model.

## 7. Comparison
Finally, lets see how the Extra Trees algorithm compares with Decision Trees and Random Forest in terms of Accuracy and Time. We will use default model parameters instead of hypertuning.

In [16]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.model_selection import cross_val_score
import time

### 7.1. Decision Tree

In [17]:
start_time = time.time()

dtc = DecisionTreeClassifier(random_state=0)
cv_score = cross_val_score(dtc, data_X, data_Y, cv=5).mean()
print("Average Cross Validation Score = ", cv_score)
print("--- %s seconds ---" % (time.time() - start_time))

Average Cross Validation Score =  0.9173730787144851
--- 0.060630083084106445 seconds ---


This algoritm is incredibly fast and took only 0.06 seconds! Reason being that it just trains a single Tree and predicts. Validation score is 91.7%

### 7.2. Random Forest

In [22]:
start_time = time.time()

rfc = RandomForestClassifier(random_state=0)
cv_score = cross_val_score(rfc, data_X, data_Y, cv=5).mean()
print("Average Cross Validation Score = ", cv_score)
print("--- %s seconds ---" % (time.time() - start_time))

Average Cross Validation Score =  0.9631113181183046
--- 1.5104308128356934 seconds ---


Random Forest took 1.52 seconds which is 25 times slower than a single Decision Tree which is obvious due to more number of estimators(trees). Validation score is 96.3%

### 7.3. Extra Trees

In [23]:
start_time = time.time()

etc = ExtraTreesClassifier(random_state=0)
cv_score = cross_val_score(etc, data_X, data_Y, cv=5).mean()
print("Average Cross Validation Score = ", cv_score)
print("--- %s seconds ---" % (time.time() - start_time))

Average Cross Validation Score =  0.9701288619779538
--- 0.9814736843109131 seconds ---


Wow! With the same default parameters, validation score of Extra Trees is very close(slightly better) than that of Random Forest i.e 97%. But the important thing to note here is the time complexity, it trains the model in somewhat 60% time as that taken by Random Forest.

## 8. Conclusion
Its hard to pin out which algorithm is better in terms of accuracy and would depend on testing. However, Extremely Randomized Trees are surely faster than Random Forest due to the random nature of picking up thresholds. Extra Trees can become a very useful algorithm if your dataset is huge and you want to quickly run a Decision Tree ensemble and check how your model performs on the dataset. Check out this link to know the exact time complexities of different machine learning algoritms(<a href="https://www.thekerneltrip.com/machine/learning/computational-complexity-learning-algorithms/">Link</a>).

### Please upvote this notebook if you found it to be useful!

Reference - Hands-On Machine Learning with Scikit-Learn, Keras and TensorFlow by Aurélien Géron